ocaml/src/blob_access.ml

```diff
@@ -0,0 +1,384 @@
+    ]
+
+class virtual blob_access (statistics : Asd_statistics.AsdStatistics.t ) =
+
+object(self)
+  val _statistics = statistics
+
+  method virtual config : config
+
+  method virtual get_blob : fnr -> int (* size of the blob *) -> bytes Lwt.t
+
+  method virtual
+      send_blob_data_to :
+        fnr -> length
+        -> (offset * length) list
+        -> Net_fd.t
+        -> unit Lwt.t
+
+  method virtual (* only used in asd_bench *)
+      push_blob_data :
+        fnr -> length
+        -> (offset * length) list
+        -> ((offset * length) -> Lwt_bytes.t -> offset -> unit Lwt.t)
+        -> unit Lwt.t
+
+  (* method virtual with_blob_fd : fnr -> (Lwt_unix.file_descr -> unit Lwt.t) -> unit Lwt.t *)
+
+  method virtual write_blob : fnr
+                              -> Asd_protocol.Blob.t
+                              -> post_write: Asd_io_scheduler.post_write
+                              -> sync_parent_dirs:bool
+                              -> unit Lwt.t
+
+  method virtual delete_blobs : fnr list -> ignore_unlink_error:bool -> unit Lwt.t
+
+  method statistics = _statistics
+  (* TODO: These should go *)
+  method virtual _get_file_dir_name_path : fnr -> bytes * bytes * bytes
+
+  method _get_file_path fnr =
+    let _, _, path = self # _get_file_dir_name_path fnr in
+    path
+end
+
+class virtual blob_dir_access statistics =
+object
+  inherit blob_access statistics
+  inherit directory_access
+end
+
+class directory_info statistics config =
+object(self)
+  inherit blob_access statistics
+  inherit default_directory_access config.files_path
+
+  method config = config
+
+  method _get_file_dir_name_path = Fnr.get_file_dir_name_path config.files_path
+
+
+  method send_blob_data_to fnr size slices nfd =
+    Lwt_extra2.with_fd
+      (if config.write_blobs
+       then self # _get_file_path fnr
+       else "/dev/zero")
+      ~flags:Lwt_unix.([O_RDONLY;])
+      ~perm:0600
+      (fun blob_fd ->
+        let blob_ufd = Lwt_unix.unix_file_descr blob_fd in
+        let () =
+          if config.use_fadvise
+          then
+            begin
+              Posix.posix_fadvise blob_ufd 0 size Posix.POSIX_FADV_RANDOM;
+              List.iter
+                (fun (offset, length) ->
+                  Posix.posix_fadvise
+                    blob_ufd
+                    offset length
+                    Posix.POSIX_FADV_WILLNEED)
+                slices
+            end
+        in
+        Lwt_list.iter_s
+          (fun (offset, length) ->
+            Net_fd.sendfile_all
+              ~fd_in:blob_fd ~offset
+              ~fd_out:nfd
+              length
+            >>= fun () ->
+            let () =
+              if config.use_fadvise
+              then
+                Posix.posix_fadvise
+                  blob_ufd
+                  0 size
+                  Posix.POSIX_FADV_DONTNEED
+            in
+            Lwt.return_unit)
+          slices
+      )
+  method push_blob_data fnr size slices f =
```

Comment:
push_blob_data and send_blob_data above seem to share a lot of code/structure, so I think it makes sense to extract out the commonalities.